### ISSUE TITLE
When timeshifting back to live the default live URL should be used

### ISSUE THREAD

**Issue by phunkyfish** [xbmc/inputstream.ffmpegdirect#17]
Currently as the switching the URL back to live causes an EOF (of some kind) and the stream ends instead we just timeshift to live -10 seconds.

**Comment by phunkyfish** [comment]
@FernetMenta could use your guidance with this as it's all built on a port of the FFmpeg demux from kodi.

For regular timeshift catchup the URL is set here: https://github.com/xbmc/inputstream.ffmpegdirect/blob/cf1dc46ab525bb68643fa5820940756a57a99b1c/src/stream/FFmpegCatchupStream.cpp#L171

I'm not sure why when I try to set the live URL the stream will just end. Possibly it's already buffered the EOF and then that is read before the live stream get's loaded. Or maybe the delay in loading the live URL is interpreted as a stream end.

Any guidance as to the right way to do this?

**Comment by FernetMenta** [comment]
I can have a look but please explain the desired behaviour of this addon.

> Possibly it's already buffered the EOF and then that is read before the live stream get's loaded

There might be something wrong with your concept. Nothing is supposed to be in a queue after a seek.

**Comment by phunkyfish** [comment]
Sure, the purpose of the addon is to support features that would be difficult to support in kodi, primarily PVR. Streams/files get opened by libavformat only.

For this use case we're supporting timeshift through catchup windows where a provider allows access to a stream at any point in time in the last few days via a URL. For the user who would be running IPTV Simple, they can navigate their EPG, channel list, previous/next all using the video player OSD. In addition they can seek to any point in the catchup window.

For this specific issue let's say there are two URLs the live URL: `http://mystream/stream.m3u8`

And the catchup URL: `http://mystream/stream.m3u8?time=13475859&duration=600`

If you seek back 10 mins or any point before live a different catchup URL will be generated. But when you jump back to live you should switch to live URL so playback continues. Regardless of the stream type the steam will always stop if jumping back to the live URL.

In the future would like to support timeshift for non catchup streams as well as stream backups for catchup too.

**Comment by FernetMenta** [comment]
Please try to put in words what you have been trying to code. Seek is not just setting a different url. What should happen in detail. When do you open and initialize the new stream url?

**Comment by phunkyfish** [comment]
Ok, so the seek window has a start time equal to the start of the catch up (e.g. 3 days ago). Each time there is a seek set operation the offset from the start time gets updated which allows a new URL to be generated. All I am doing is setting the new URL. There is no open/initialise of the new URL. 

I had crudely assumed this would automatically just open the new URL and play (which it does). And as it worked I didn't try to take it any further ;)

But clearly this has its faults.

**Comment by FernetMenta** [comment]
Talk is about m3u8 so I would assume that ffmpeg's hls demuxer is used here. Correct?

**Comment by phunkyfish** [comment]
Primarily yes. But it will work with TS streams and unencrypted DASH that I have tested so far.

**Comment by FernetMenta** [comment]
> I had crudely assumed this would automatically just open the new URL and play (which it does)

Changing the IO context without closing the format context violates the programming model of ffmpeg's API. ffmpeg stores information about the source it reads from. You cannot simply change the source on the fly. If you want to change the input, you have to close and reopen the format context too.

**Comment by phunkyfish** [comment]
Ok. So would this mean calling Dispose() and Open() straight after?

Or is the sequence of steps different if I don’t want the inputstream to close when doing this?

Sorry if the questions are a bit basic. I’m learning 😉

**Comment by phunkyfish** [comment]
@FernetMenta, thanks for the guidance so far. So I now have this working correctly for the majority of streams. Correctly resetting the demuxer when changing stream URLs.

But I could use your insight again.

I have one set of streams where there are issues. I can play the catchup stream from the start no problem, or the live stream no problem and there is no issue until I switch from one to the other.

Here is a log when started with the catchup stream (working): https://gist.github.com/phunkyfish/7931b21c55fa4633b603a3d147584a05

Here is a log when started with the live stream (working): https://gist.github.com/phunkyfish/04a48407e9f5bbccffe17fa7953770fd

Finally here is the log that freezes when the switch from catchup (working) to live (freezing) occurs:
https://gist.github.com/phunkyfish/fd8f8edb1af4d8198427d72fd93fc845

I get a lot of these just not not sure what they mean: 

```
 ffmpeg[(nil)X]: [h264] No start code is found.
 ffmpeg[(nil)X]: [h264] Error splitting the input into NAL units.
```

**Comment by phunkyfish** [comment]
Fixed in v1.9.0
